test/hotspot/jtreg/serviceability/jvmti/vthread/ToggleNotifyJvmtiTest/ToggleNotifyJvmtiTest.java

```diff
@@ -0,0 +1,178 @@
+    // We will temporarily set a breakpoint on this method when the thread should be suspended.
+    // If we hit the breakpoint, then something is wrong.
+    public void breakpointCheck() {
+        return;
+    }
+
+    // run thread continuously
+    public void run() {
+        // run in a loop
+        threadReady = true;
+        int i = 0;
+        int n = 1000;
+        while (!shouldFinish) {
+            breakpointCheck();
+            if (n <= 0) {
+                n = 1000;
+                ToggleNotifyJvmtiTest.sleep(1);
+            }
+            if (i > n) {
+                i = 0;
+                n = n - 1;
+            }
+            i = i + 1;
+        }
+    }
+
+    // ensure thread is ready
+    public void ensureReady() {
+        try {
+            while (!threadReady) {
+                sleep(1);
+            }
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Interruption while preparing tested thread: \n\t" + e);
+        }
+    }
+
+    public void letFinish() {
+        shouldFinish = true;
+    }
+}
+
+public class ToggleNotifyJvmtiTest {
+    private static final int VTHREADS_CNT = 60;
+    private static final String AGENT_LIB = "ToggleNotifyJvmtiTest";
+    private static final WhiteBox WB = WhiteBox.getWhiteBox();
+
+    private static native boolean IsAgentStarted();
+    private static native int VirtualThreadStartedCount();
+
+    static void log(String str) { System.out.println(str); }
+
+    static public void sleep(long millis) {
+        try {
+            Thread.sleep(millis);
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Interruption in TestedThread.sleep: \n\t" + e);
+        }
+    }
+
+    static TestedThread[] threads = new TestedThread[VTHREADS_CNT];
+    static Thread vts[] = new Thread[VTHREADS_CNT];
+
+    static private synchronized void startThread(int i) {
+        String name = "TestedThread" + i;
+        TestedThread thread = new TestedThread(name);
+        vts[i] = Thread.ofVirtual().name(name).start(thread);
+        thread.ensureReady();
+        threads[i] = thread;
+        log("## Java: started vthread: " + name);
+    }
+
+    static private void startThreads() {
+        log("\n## Java: Starting vthreads");
+        for (int i = 0; i < VTHREADS_CNT; i++) {
+            startThread(i);
+        }
+    }
+
+    static private synchronized void finishThreads() {
+        log("\n## Java: runIt: Finishing vthreads");
+        try {
+            for (int i = 0; i < VTHREADS_CNT; i++) {
+                TestedThread thread = threads[i];
+                if (thread == null) {
+                    break;
+                }
+                thread.letFinish();
+                vts[i].join();
+            }
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    static private void setVirtualThreadsNotifyJvmtiMode(int iter, boolean enable) {
+        sleep(5);
```

Comment:
Why is this needed?

Reply:
It is needed to balance enabling/disabling notifyJvmti mode with the ThreadStart/VirtualThreadStart events.
Otherwise, many mode switches can be observed without any events which is not interesting.

Reply:
Shouldn't that be the caller's responsibility? Including a comment would be helpful.

Reply:
Okay, moved to the caller. In fact, I've refactored the test even more for safety purposes. Will describe it separately.